File: cheatsheet/编程笔记/_util/java/FileUtil.java
/**
 * <P> Title: 公用類別                                      </P>
 * <P> Description: 檔案操作工具                            </P>
 * <P> Copyright: Copyright (c) 2010/08/20                  </P>
 * <P> Company:Everunion Tech. Ltd.                         </P>
 */

package com.util;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

/**
 * 檔案操作工具
 * @author Holer W. L. Feng
 * @version 0.1
 */
public class FileUtil
{
    /**
     * 創建此抽象路徑名指定的目錄，包括創建必需但不存在的父目錄
     * @param dir 指定的目錄
     * @return 當且僅當已創建該目錄以及所有必需的父目錄時，傳回 true；否則傳回 false 
     */
    public static boolean createDir(String dir)
    {
        boolean flag = false;
        File file = new File(dir);
        // 目錄不存在或者不是目錄時，創建
        if ( !file.exists() && !file.isDirectory() )
        {
            flag = file.mkdirs();
        }
        return flag;
    }
    

    /**
     * 建立檔案；如果之前已經有此檔案，則取代它；
     * @param fileName 檔案的檔案名；如果檔案的目錄不存在，則創建目錄
     * @param message 需要寫入到屬性檔案的內容
     * @param csn 建立此屬性檔案的字元集
     * @return boolean 是否操作成功：true則成功，false則操作不成功
     */
    public static boolean createFile ( String fileName, String message )
    {
        // 防呆
        if ( fileName == null )
            return false;
        PrintWriter pw = null;
        // 檔案的目錄
        File parentFile = new File(fileName).getParentFile();
        // 如果檔案的目錄不存在，則先創建目錄
        if ( !parentFile.exists() )
            parentFile.mkdirs();
        try
        {
            pw = new PrintWriter(fileName, "UTF-8");
            // 寫出內容
            if ( message != null && !"".equals(message) )
                pw.print(message);
        }
        catch ( FileNotFoundException e )
        {
            System.out.println("FileUtil.createFile:" + e.toString());
            return false;
        }
        catch ( UnsupportedEncodingException e )
        {
            System.out.println("FileUtil.createFile:" + e.toString());
            return false;
        }
        finally
        {
            // 清空緩存
            pw.flush();
            // 關閉輸出串流
            pw.close();
        }
        return true;
    }
    

    public static boolean copyDir(String s, String s1) throws IOException
    {
        File file = new File(s);
        File file1 = new File(s1);
        if ( file.exists() )
        {
            if ( !file1.isDirectory() )
            {
                boolean flag = file1.mkdirs();
                if ( !flag )
                    return false;
            }
            String as[] = file.list();
            for ( int i = 0; i < as.length; i++ )
            {
                File file2 = new File(s + File.separator + as[i]);
                if ( file2.isFile() )
                    copyFile(s + File.separator + as[i], s1 + File.separator + as[i]);
                else if ( file2.isDirectory() && as[i].compareTo(".") != 0
                        && as[i].compareTo("..") != 0 )
                {
                    boolean flag1 = copyDir(s + File.separator + as[i], s1
                            + File.separator + as[i]);
                    if ( !flag1 )
                        return false;
                }
            }
        }
        return true;
    }
    

    public static void copyFile(String s, String s1) throws IOException
    {
        FileInputStream fileinputstream = null;
        FileOutputStream fileoutputstream = null;
        byte abyte0[] = new byte[8192];
        try
        {
            fileinputstream = new FileInputStream(s);
            fileoutputstream = new FileOutputStream(s1);
            int i;
            while ( (i = fileinputstream.read(abyte0)) != -1 )
                fileoutputstream.write(abyte0, 0, i);
        }
        finally
        {
            try
            {
                if ( fileinputstream != null )
                    fileinputstream.close();
                if ( fileoutputstream != null )
                    fileoutputstream.close();
            }
            catch ( IOException ioexception )
            {
            }
        }
    }
    

    public static boolean delFolder(String s)
    {
        File file = new File(s);
        if ( file.exists() && file.isDirectory() )
        {
            File afile[] = file.listFiles();
            for ( int i = 0; i < afile.length; i++ )
            {
                if ( afile[i].isFile() )
                {
                    if ( afile[i].exists() )
                        afile[i].delete();
                    continue;
                }
                if ( afile[i].isDirectory() )
                    delFolder(afile[i].toString());
            }
            file.delete();
        }
        file = null;
        return true;
    }
    

    public static final void delFile(String s) throws Exception
    {
        File file = new File(s);
        if ( file.exists() )
        {
            file.delete();
        }
    }
    

    public static boolean moveDir(String s, String s1) throws IOException
    {
        if ( copyDir(s, s1) )
            return delFolder(s);
        else
            return false;
    }
    

    public static String[] listFile(String s)
    {
        ArrayList<String> vector = new ArrayList<String>();
        File file = new File(s);
        if ( file.exists() )
        {
            File afile[] = file.listFiles();
            for ( int i = 0; i < afile.length; i++ )
            {
                if ( afile[i].isFile() )
                    vector.add(afile[i].getName());
            }
        }
        return (String[])vector.toArray(new String[0]);
    }
    

    public static String[] listFolder(String s)
    {
        ArrayList<String> vector = new ArrayList<String>();
        File file = new File(s);
        if ( file.exists() )
        {
            File afile[] = file.listFiles();
            for ( int i = 0; i < afile.length; i++ )
            {
                if ( afile[i].isDirectory() )
                    vector.add(afile[i].getName());
            }
            
        }
        return (String[])vector.toArray(new String[0]);
    }
    

    public static String readFile2String(String name) throws Exception
    {
        StringBuffer sb = new StringBuffer();
        FileInputStream fis = null;
        InputStreamReader isr = null;
        BufferedReader br = null;
        try
        {
            fis = new FileInputStream(name);
            isr = new InputStreamReader(fis, "utf-8");
            br = new BufferedReader(isr);
            String data = null;
            while ( (data = br.readLine()) != null )
            {
                sb.append(data);
                sb.append("\n");
            }
        }
        catch ( Exception e )
        {
            System.out.println("FileUtil.readFile2String:" + e.toString());
        }
        finally
        {
            if ( br != null )
            {
                br.close();
                br = null;
            }
            if ( isr != null )
            {
                isr.close();
                isr = null;
            }
            if ( fis != null )
            {
                fis.close();
                fis = null;
            }
        }
        return sb.toString();
    }
    

    /**
     * 在原檔案名的後面加上時間(以免名稱衡突)
     * @param input 檔案名
     * @return String 處理後的檔案名
     */
    public static String getFileNameByTime(String input)
    {
        // 取出副檔名
        int index = input.indexOf(".");
        Date dt = new Date();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMddHHmmssSSS");
        // 原檔案名(無副檔名) + 時間 + 副檔名
        return input.substring(0, index) + sdf.format(dt) + input.substring(index);
    }
    
}

File: cheatsheet/编程笔记/_util/c#/FileUtil.cs
/**
 * <P> Title: 工具类                        </P>
 * <P> Description: 工具类                  </P>
 * <P> Copyright: Copyright (c) 2010-07-13  </P>
 * <P> Company:Everunion Tech. Ltd.         </P>
 * @author <A href="daillow@gmail.com">Holer</A>
 */
using System;
using System.Collections;
using System.Configuration;
using System.IO;
using System.Text;

namespace Com.Util
{
    public class FileUtil
    {

        /// <summary>
        /// 创建文件
        /// </summary>
        /// <param name="fileName">文件名</param>
        /// <param name="strMsg">文件里需要写入的数据</param>
        /// <param name="overwritten">如果文件已经存在,是否覆盖;true则覆盖旧档,false则保留旧文件而追加数据到文件后面</param>
        public static void WriteFile(string fileName, string strMsg, bool overwritten)
        {
            FileInfo file = new FileInfo(fileName);
            //如果文件路径不存在,先创建文件夹
            if ( !file.Directory.Exists )
            {
                file.Directory.Create();
            }
            //如果文件已经存在;overwritten 为true则覆盖旧档,false则保留旧文件而追加数据到文件后面
            FileMode fileMode = overwritten ? FileMode.Create : FileMode.Append;
            FileStream fs = new FileStream(fileName, fileMode);
            StreamWriter sw = new StreamWriter(fs, Encoding.UTF8);
            //写入资料
            sw.WriteLine(strMsg);
            sw.Flush();
            //关闭流
            sw.Close();
            fs.Close();
            //销毁对象
            sw.Dispose();
            fs.Dispose();
        }


        /// <summary>
        /// 创建文件; 如果文件已经存在,则覆盖旧档
        /// </summary>
        /// <param name="fileName">文件名</param>
        /// <param name="strMsg">文件里需要写入的数据</param>
        public static void WriteFile(string fileName, string strMsg)
        {
            WriteFile( fileName, strMsg, true );
        }


        /// <summary>
        /// 删除目录下的所有文件
        /// </summary>
        /// <param name="Dir">文件目录(不存在时创建)</param>
        public static void DeleteDirFiles(string Dir)
        {
            //目录不存在
            if (!System.IO.Directory.Exists(Dir))
            {
                System.IO.Directory.CreateDirectory(Dir);
            }
            //目录存在
            else
            {
                //文件数组
                string[] Files = System.IO.Directory.GetFiles(Dir);
                foreach (string file in Files)
                {
                    //删除
                    System.IO.File.Delete(file);
                }
            }
        }


        /// <summary>
        /// 删除目录下的所有文件
        /// </summary>
        /// <param name="Dir">文件目录(不存在时创建)</param>
        public static void DeleteFile(string fileName)
        {
            try
            {
                System.IO.File.Delete(fileName);
            }
            catch
            {
            }
        }


        /// <summary> (此函数直接复制过来，未测试)
        /// 获取图片，并保存到本地
        /// </summary>
        /// <param name="imgUrl">图片地址</param>
        /// <param name="path">保存路径</param>
        /// <param name="fileName">图片名</param>
        /// <returns>图片保存成功返回true，否则返回false</returns>
        public static bool SaveImageFromWeb(string imgUrl, string path, string fileName)
        {
            // 未指定保存文件的路径
            if (string.IsNullOrEmpty(path))
                throw new Exception("未指定保存文件的路径");
            // 未指定图片的地址
            if (string.IsNullOrEmpty(imgUrl))
                throw new Exception("未指定保存文件的路径");

            string imgName = imgUrl.Substring(imgUrl.LastIndexOf("/") + 1);

            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(imgUrl);
            request.UserAgent = "Mozilla/6.0 (MSIE 6.0; Windows NT 5.1; Natas.Robot)";
            request.Timeout = 3000;

            WebResponse response = request.GetResponse();
            Stream stream = response.GetResponseStream();

            if (response.ContentType.ToLower().StartsWith("image/"))
            {
                byte[] arrayByte = new byte[1024];
                int imgLong = (int)response.ContentLength;
                int l = 0;

                // 图片类型
                string defaultType = "." + response.ContentType.Split('/')[1];
                string[] imgTypes = new string[] { ".jpg", ".jpeg", ".png", ".gif", ".bmp" };
                string imgType = imgUrl.Substring(imgUrl.LastIndexOf("."));
                foreach (string it in imgTypes)
                {
                    if (imgType.ToLower().Equals(it))
                        break;
                    // 如果不属于上诉的类型，使用默认的
                    if (it.Equals(".bmp"))
                        imgType = defaultType;
                }

                //如果文件名没有传进来，使用图片的
                if (string.IsNullOrEmpty(fileName))
                    fileName = imgName;

                FileStream fso = new FileStream(path + fileName + imgType, FileMode.Create);
                while (l < imgLong)
                {
                    int i = stream.Read(arrayByte, 0, 1024);
                    fso.Write(arrayByte, 0, i);
                    l += i;
                }

                fso.Close();
                stream.Close();
                response.Close();

                return true;
            }
            else
            {
                return false;
            }
        }



    }
}
